python/lsst/daf/butler/registry/_registry.py

```diff
@@ -820,12 +826,103 @@ def insertDatasets(self, datasetType: Union[DatasetType, str], dataIds: Iterable
+            Datasets to be inserted. All `DatasetRef` instances must have
+            identical ``datasetType`` and ``run`` attributes. ``run``
+            attribute can be `None` and defaults to ``self.defaults.run``.
+            Datasets can specify ``id`` attribute which will be used for
+            inserted datasets. All dataset IDs must have the same type
+            (`int` or `uuid.UUID`), if type of dataset IDs does not match
+            configured backend then IDs will be ignored and new IDs will be
+            generated by backend.
+        idGenerationMode : `DatasetIdGenEnum`, optional
+            Specifies option for generating dataset IDs when IDs are not
+            provided or their type does not match backend type. By default
+            unique IDs are generated for each inserted dataset.
+
+        Returns
+        -------
+        refs : `list` of `DatasetRef`
+            Resolved `DatasetRef` instances for all given data IDs (in the same
+            order). If any of ``datasets`` has an ID which already exists in
+            the database then it will not be inserted or updated, but a
+            resolved `DatasetRef` will be returned for it in any case.
+
+
+        Raises
+        ------
+        TypeError
+            Raised if ``run`` is `None` and ``self.defaults.run`` is `None`.
+        ConflictingDefinitionError
+            If a dataset with the same dataset type and data ID as one of those
+            given already exists in ``run``.
+        MissingCollectionError
+            Raised if ``run`` does not exist in the registry.
+        """
+        datasets = list(datasets)
+        if not datasets:
+            # nothing to do
+            return []
+
+        # find dataset type
+        datasetTypes = set(dataset.datasetType for dataset in datasets)
+        if len(datasetTypes) != 1:
+            raise ValueError(f"Multiple dataset types in input datasets: {datasetTypes}")
+        datasetType = datasetTypes.pop()
+
+        # get storage handler for this dataset type
+        storage = self._managers.datasets.find(datasetType.name)
+        if storage is None:
+            raise LookupError(f"DatasetType '{datasetType}' has not been registered.")
+
+        # find run name
+        runs = set(dataset.run for dataset in datasets)
+        if len(runs) != 1:
+            raise ValueError(f"Multiple run names in input datasets: {runs}")
+        run = runs.pop()
+        if run is None:
+            if self.defaults.run is None:
+                raise TypeError("No run provided to ingestDatasets, "
+                                "and no default from registry construction.")
+            run = self.defaults.run
+
+        runRecord = self._managers.collections.find(run)
+        if runRecord.type is not CollectionType.RUN:
+            raise TypeError(f"Given collection is of type {runRecord.type.name}; RUN collection required.")
+        assert isinstance(runRecord, RunRecord)
```

Comment:
This was copied verbatim from `insertDatasets()`, I'll see if it can be removed.

Reply:
OK, this is indeed for mypy only, `runRecord` cannot be `None` here but it can be any other subclass of `CollectionRecord`. Checking that `runRecord.type` is `RUN` works to constrain the type but mypy still needs convincing.